### PR TITLE
[DeepSeek] Part 2 of Refactor to Align With vLLM Modeling Conventions

### PR DESCRIPTION
**Pull request by jrplatin**
# Description

Builds upon the first part of the DeepSeek refactor as outlined in https://github.com/vllm-project/tpu-inference/pull/1581.   Copying over context from there:

Currently, models in the JAX/Flax path in [uLLM](https://github.com/vllm-project/tpu-inference/tree/main/tpu_inference/models/jax) follow two different implementation designs:

Qwen2/Qwen2.5/Qwen3/Llama3 implement each layer (e.g. Attention, DecoderLayer, MLP, etc.) per-model. That is, Llama3 has a Llama3Attention, Qwen2 has a Qwen2Attention and so on
Llama4/DeepSeek/GPT-OSS, however, follow a more hybrid approach where some layers are model-specific (e.g. DeepSeekRouter) and others are not (e.g. TransformerBlock and SharedExpertsTransformerBlock)
For comparison, Qwen/DeepSeek in vLLM follows the [first](https://github.com/vllm-project/vllm/blob/main/vllm/model_executor/models/deepseek_v2.py) paradigm.

Given TPU Inference’s role as a plug-in to vLLM, we should align with their conventions.  Specifically, we define DeepSeek-specific attention classes: one for MHA and one for MLA, which also makes the code a bit easier to navigate.

# Tests

Reverted 2D-TP change to test accuracy (since those changes are causing issues with the Megablocks kernel) and everything looks on par:

`

MLA:

```
feature branch
USE_MEGABLOCKS=1 NEW_MODEL_DESIGN=True TPU_BACKEND_TYPE=jax  vllm serve --model=jrplatin/DeepSeek-R1-1D-Subchannel-256-FP8 --max-model-len=5120 --tensor-parallel-size 8 --max-num-seqs 16 --max-num-batched-tokens 128 --no-enable-prefix-caching --disable-log-requests --gpu-memory-utilization 0.95 --additional_config='{"sharding": {"sharding_strategy": {"attention_data_parallelism": 8, "enable_dp_attention": true}}}' 
============ Serving Benchmark Result ============
Successful requests:                     1000      
Benchmark duration (s):                  103.98    
Total input tokens:                      162129    
Total generated tokens:                  2000      
Request throughput (req/s):              9.62      
Output token throughput (tok/s):         19.23     
Total token throughput (tok/s):          1578.49   
---------------Time to First Token----------------
Mean TTFT (ms):                          59756.32  
Median TTFT (ms):                        62716.96  
P99 TTFT (ms):                           102917.93 
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          631.33    
Median TPOT (ms):                        634.56    
P99 TPOT (ms):                           638.76    
---------------Inter-token Latency----------------
Mean ITL (ms):                           631.33    
Median ITL (ms):                         634.56    
P99 ITL (ms):                            638.76    
----------------End-to-end Latency----------------
Mean E2EL (ms):                          60387.64  
Median E2EL (ms):                        63352.33  
P99 E2EL (ms):                           103578.91 
==================================================
Evaluating MMLU...
[nltk_data] Downloading package punkt to
[nltk_data]     /home/jacobplatin_google_com/nltk_data...
[nltk_data]   Package punkt is already up-to-date!
[nltk_data] Downloading package punkt_tab to
[nltk_data]     /home/jacobplatin_google_com/nltk_data...
[nltk_data]   Package punkt_tab is already up-to-date!

Results

{'accuracy': 0.849, 'gen_num': 1000}



`main`

============ Serving Benchmark Result ============
Successful requests:                     1000      
Benchmark duration (s):                  103.69    
Total input tokens:                      162129    
Total generated tokens:                  2000      
Request throughput (req/s):              9.64      
Output token throughput (tok/s):         19.29     
Total token throughput (tok/s):          1582.82   
---------------Time to First Token----------------
Mean TTFT (ms):                          59583.17  
Median TTFT (ms):                        62697.24  
P99 TTFT (ms):                           102259.02 
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          631.19    
Median TPOT (ms):                        634.90    
P99 TPOT (ms):                           637.89    
---------------Inter-token Latency----------------
Mean ITL (ms):                           631.19    
Median ITL (ms):                         634.90    
P99 ITL (ms):                            637.89    
----------------End-to-end Latency----------------
Mean E2EL (ms):                          60214.37  
Median E2EL (ms):                        63333.81  
P99 E2EL (ms):                           102892.47 
==================================================
Evaluating MMLU...
[nltk_data] Downloading package punkt to
[nltk_data]     /home/jacobplatin_google_com/nltk_data...
[nltk_data]   Package punkt is already up-to-date!
[nltk_data] Downloading package punkt_tab to
[nltk_data]     /home/jacobplatin_google_com/nltk_data...
[nltk_data]   Package punkt_tab is already up-to-date!

Results

{'accuracy': 0.846, 'gen_num': 1000}

```

MHA:

```
USE_MEGABLOCKS=1 VLLM_MLA_DISABLE=1 NEW_MODEL_DESIGN=1 TPU_BACKEND_TYPE=jax vllm serve --model=jrplatin/DeepSeek-R1-1D-Subchannel-256-FP8  --max-model-len=5120 --max-num-batched-tokens 128 --max-num-seqs=128 --no-enable-prefix-caching --disable-log-requests --gpu-memory-utilization 0.95 --tensor-parallel-size 8 --additional_config='{"sharding": {"sharding_strategy": {"expert_parallelism": 1, "tensor_parallelism": 8}}}' --kv-cache-dtype=fp8  


feature branch:

Successful requests:                     1000      
Benchmark duration (s):                  327.73    
Total input tokens:                      162129    
Total generated tokens:                  2000      
Request throughput (req/s):              3.05      
Output token throughput (tok/s):         6.10      
Total token throughput (tok/s):          500.80    
---------------Time to First Token----------------
Mean TTFT (ms):                          185023.96 
Median TTFT (ms):                        194969.32 
P99 TTFT (ms):                           323800.51 
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          255.75    
Median TPOT (ms):                        256.62    
P99 TPOT (ms):                           262.20    
---------------Inter-token Latency----------------
Mean ITL (ms):                           255.75    
Median ITL (ms):                         256.62    
P99 ITL (ms):                            262.20    
----------------End-to-end Latency----------------
Mean E2EL (ms):                          185279.71 
Median E2EL (ms):                        195222.30 
P99 E2EL (ms):                           324061.20 
==================================================
Evaluating MMLU...
[nltk_data] Downloading package punkt to
[nltk_data]     /home/jacobplatin_google_com/nltk_data...
[nltk_data]   Package punkt is already up-to-date!
[nltk_data] Downloading package punkt_tab to
[nltk_data]     /home/jacobplatin_google_com/nltk_data...
[nltk_data]   Package punkt_tab is already up-to-date!

Results

{'accuracy': 0.849, 'gen_num': 1000}



`main` before changes:
============ Serving Benchmark Result ============
Successful requests:                     1000      
Benchmark duration (s):                  327.56    
Total input tokens:                      162129    
Total generated tokens:                  2000      
Request throughput (req/s):              3.05      
Output token throughput (tok/s):         6.11      
Total token throughput (tok/s):          501.07    
---------------Time to First Token----------------
Mean TTFT (ms):                          184896.43 
Median TTFT (ms):                        194766.39 
P99 TTFT (ms):                           323374.19 
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          255.57    
Median TPOT (ms):                        256.58    
P99 TPOT (ms):                           262.02    
---------------Inter-token Latency----------------
Mean ITL (ms):                           255.57    
Median ITL (ms):                         256.58    
P99 ITL (ms):                            262.02    
----------------End-to-end Latency----------------
Mean E2EL (ms):                          185152.00 
Median E2EL (ms):                        195019.17 
P99 E2EL (ms):                           323626.62 
==================================================
Evaluating MMLU...
[nltk_data] Downloading package punkt to
[nltk_data]     /home/jacobplatin_google_com/nltk_data...
[nltk_data]   Package punkt is already up-to-date!
[nltk_data] Downloading package punkt_tab to
[nltk_data]     /home/jacobplatin_google_com/nltk_data...
[nltk_data]   Package punkt_tab is already up-to-date!

Results

{'accuracy': 0.851, 'gen_num': 1000}


```

# Checklist

Before submitting this PR, please make sure:
- I have performed a self-review of my code.
- I have necessary comments in my code, particularly in hard-to-understand areas.
- I have made or will make corresponding changes to any relevant documentation.

### CODE REVIEW

File: tests/models/jax/test_deepseek_v3.py [hunk 1]
```diff
@@ -21,11 +22,19 @@
...
 # Assuming the model file is named deepseek_v3.py
-from tpu_inference.models.jax.deepseek_v3 import (DeepSeekV3, DeepSeekV3Router,
+import tpu_inference.kernels.mla.v1.kernel as mla
+from tpu_inference.layers.common.attention_interface import get_kv_cache_shape
+from tpu_inference.layers.common.attention_metadata import AttentionMetadata
+from tpu_inference.layers.common.sharding import ShardingAxisName
+from tpu_inference.models.jax.deepseek_v3 import (DeepSeekV3,
+                                                  DeepseekV3Attention,
+                                                  DeepseekV3MLA,
+                                                  DeepSeekV3Router,
```
Comment (patemotter) [inline tests/models/jax/test_deepseek_v3.py]: Nit: number of new lines seems a bit excessive

File: tests/scripts/run_rpa_v3_tests.sh [hunk 2]
```diff
@@ -33,3 +32,4 @@ done
 
 # Run all tests in a single pytest command
 pytest "${FULL_PATHS[@]}"
+pytest "$TPU_INFERENCE_DIR/tpu-inference/tests/models/jax/test_deepseek_v3.py" -k "TestDeepseekV3Attention"
```
Comment (patemotter) [inline tests/scripts/run_rpa_v3_tests.sh:35]: SuperNit: Change `# Run all tests in a single pytest command` since that is not longer accurate. 

`FULL_PATHS` also seems a bit overkill now, but we can keep it if you think it's beneficial.

File: tpu_inference/models/jax/deepseek_v3.py [hunk 3]
```diff
@@ -56,6 +65,404 @@
...
+    hidden_size: int
+    num_attention_heads: int
+    num_key_value_heads: int
+    head_dim: int
+    rope_theta: float
+    rope_scaling: dict[str, Any]
+    dtype: jnp.dtype
+    kv_cache_dtype: str
+    mesh: Mesh
+
```
Comment (patemotter) [inline tpu_inference/models/jax/deepseek_v3.py:84]: Nit: Adding newlines definitely helps with dividing up the data logically. Could we also add a single comment for each group to group them?

File: tpu_inference/models/jax/deepseek_v3.py [hunk 4]
```diff
@@ -56,6 +65,404 @@
...
+    def compute_kv_projection(self, *args):
+        raise NotImplementedError
+
+    def compute_attention(self, *args):
+        raise NotImplementedError
+
+    def process_output(self, outputs_TNH):
+        return outputs_TNH
+
+    def __call__(self, x, is_prefill, kv_cache: KVCache,
```
Comment (patemotter) [inline tpu_inference/models/jax/deepseek_v3.py]: Nit: missing datatype hints for `x` and `is_prefill`.

Comment (patemotter) [reply tpu_inference/models/jax/deepseek_v3.py]: Looking at more of the code in DeepseekV3Attention there's a lot of missing datatypes in functions. If you want to get a good samaritan award you could add them throughout. At the very least a TODO mentioning this could be helpful.

**Review by patemotter — APPROVED** [review]
Overall it LGTM. Some small comments, but nothing regarding the core logic.

File: tpu_inference/models/jax/deepseek_v3.py (hunk 4 above)
Comment (jrplatin) [reply tpu_inference/models/jax/deepseek_v3.py]: Ah yep I'm usually good at this but missed it, thanks!

File: tests/scripts/run_rpa_v3_tests.sh (hunk 2 above)
Comment (jrplatin) [reply tests/scripts/run_rpa_v3_tests.sh:35]: Good catch! I foresee expanding those tests, so might as well keep that logic for now I think

File: tpu_inference/models/jax/deepseek_v3.py (hunk 4 above)
Comment (jrplatin) [reply tpu_inference/models/jax/deepseek_v3.py]: added!

File: tpu_inference/models/jax/deepseek_v3.py (hunk 3 above)
Comment (jrplatin) [reply tpu_inference/models/jax/deepseek_v3.py:84]: Added!

**Review by bzgoogle — APPROVED** [review]
